<launch>
  # 로봇이 주행한 환경을 2차원 지도로 기록한다.
  # 로봇: Turtlebot3 waffle(beta)
  # 제어방식: 속도제어, 키보드 awsdx
  # 센서: RPLidar A3(ttyUSB0)
  # SLAM 패키지: GMapping
  # 지도저장 명령어: rosrun map_server map_saver -f ~/map

  <!-- robot -->
  <include file="$(find shared_control)/launch/turtlebot.launch"/>

  <!-- control -->
  <node pkg="turtlebot3_teleop" type="turtlebot3_teleop_key"
        name="turtlebot3_teleop_keyboard"  output="screen"/>

  <!-- SLAM -->
  <node pkg="gmapping" type="slam_gmapping" name="turtlebot3_slam_gmapping">
    <param name="base_frame" value="base_footprint"/>
    <param name="odom_frame" value="odom"/>
    <param name="map_frame"  value="map"/>
    <param name="map_update_interval" value="2.0"/>
    <param name="maxUrange" value="10.0"/>
    <param name="maxRange" value="25.0"/>
    <param name="sigma" value="0.05"/>
    <param name="kernelSize" value="1"/>
    <param name="lstep" value="0.05"/>
    <param name="astep" value="0.05"/>
    <param name="iterations" value="5"/>
    <param name="lsigma" value="0.075"/>
    <param name="ogain" value="3.0"/>
    <param name="lskip" value="0"/>
    <param name="minimumScore" value="50"/>
    <param name="srr" value="0.1"/>
    <param name="srt" value="0.2"/>
    <param name="str" value="0.1"/>
    <param name="stt" value="0.2"/>
    <param name="linearUpdate" value="1.0"/>
    <param name="angularUpdate" value="0.2"/>
    <param name="temporalUpdate" value="0.5"/>
    <param name="resampleThreshold" value="0.5"/>
    <param name="particles" value="100"/>
    <param name="xmin" value="-30.0"/>
    <param name="ymin" value="-30.0"/>
    <param name="xmax" value="30.0"/>
    <param name="ymax" value="30.0"/>
    <param name="delta" value="0.05"/>
    <param name="llsamplerange" value="0.01"/>
    <param name="llsamplestep" value="0.01"/>
    <param name="lasamplerange" value="0.005"/>
    <param name="lasamplestep" value="0.005"/>
  </node>

  <!-- visualize -->
  <node pkg="rviz" type="rviz" name="rviz" respawn="true"
        args="-d $(find shared_control)/rviz/slam.rviz"/>
</launch>
